File: src/metrosucks.exs
defmodule MetroSucks do

	@alamein 	%{ name: 'Alemein', 	  stations: ['Flinders Street','Richmond','East Richmond','Burnley','Hawthorn','Glenferrie'] }
	@sandringham 	%{ name: 'Sandringham',   stations: ['Southern Cross','Richmond','South Yarra','Prahran','Windsor'] }
	@glen_waverley 	%{ name: 'Glen Waverley', stations: ['Flagstaff','Melbourne Central','Parliament','Richmond','Kooyong','Tooronga'] }
	
	@all_lines 		[@alamein, @sandringham, @glen_waverley]
	@interchange    'Richmond'

	def commute origin, destination do
		if different_line origin, destination do
			Enum.concat( map_journey(origin, 	   @interchange), 
	 					 map_journey(@interchange, destination) )
		else
			map_journey origin, destination
		end
		|> Enum.uniq
		|> Enum.join( " --> " )
		|> IO.puts
	end

	defp different_line( @interchange, _), do: false
	defp different_line( _, @interchange), do: false
	defp different_line origin, destination do
		what_line(origin) != what_line(destination)
	end

	defp what_line (@interchange), do: :dunno
	defp what_line station do
		Enum.find(@all_lines, fn line -> Enum.any?(line.stations, fn s -> s === station end )  end)
	end

	defp map_journey origin, destination do
		_ = origin
		_ = destination
		['Windsor', 'Prahran', 'South Yarra']
	end

end

MetroSucks.commute 'Glenferrie', 'Southern Cross' 

File: src/metrosucks.js

var alamein = ['Flinders Street','Richmond','East Richmond','Burnley','Hawthorn','Glenferrie'];
var sandringham = ['Southern Cross','Richmond','South Yarra','Prahran','Windsor'];
var glenWaverley =  ['Flagstaff','Melbourne Central','Parliament','Richmond','Kooyong','Tooronga'];
var allLines = [alamein,sandringham,glenWaverley];

var differentLine = function(origin,destination) {
 if( origin === interchange() || destination === interchange() ) {
   return false;
 }
 return (whatLine(origin)).indexOf(destination) === -1 
}

var interchange = function() {
 return 'Richmond';
}

var whatLine = function(station) {
 if( station === interchange() ) {
   return undefined;
 }  

 for (var i = 0;i < allLines.length;i++) {
   var lineLength = allLines[i].length;
   for (var j = 0; j < lineLength; j++ ) {
     if (allLines[i][j] === station) {
        return allLines[i]; 
     }
   }
 }
}

var mapJourney = function(line,origin,destination) {
 var start = line.indexOf(origin);
 var end = line.indexOf(destination);

 if (end > start) {
  return line.slice(start,end+1);
 } 
 return line.slice(end,start+1).reverse();
} 


var origin = 'Glenferrie';
// var origin = 'Richmond';
var destination = 'Flagstaff';

var completeJourney = [];
if (differentLine(origin,destination)) {
 var changeAt = interchange();
 var journey1 = mapJourney(whatLine(origin),origin,changeAt);
 var journey2 = mapJourney(whatLine(destination),changeAt,destination);

 journey1.pop();
 completeJourney = journey1.concat(journey2);
} else {
 
 if( whatLine(origin) === undefined ) {
   completeJourney = mapJourney(whatLine(destination), origin, destination);
 } else {
   completeJourney = mapJourney(whatLine(origin), origin, destination);
 }
}

console.log("------------------------------------");
console.log(" >>>  Welcome to Metro™ <<<");
console.log("Starting at: " + origin);
console.log("   Going to: " + destination);
console.log("------------------------------------");

console.log( completeJourney.join(' --> ') );

console.log("Your journey will go for " + completeJourney.length + " stops")
if( differentLine(origin, destination) ) {
	console.log("You will need to change at " + interchange() + " station");
}


